bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/EnOceanHandlerFactory.java

```diff
@@ -55,6 +57,7 @@
     private Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
 
     private SerialPortManager serialPortManager;
+    private ItemChannelLinkRegistry itemChannelLinkRegistry;
```

Comment:
We are on a more modern OSGi now. Can you please use `@Reference` on the field if you do not use the methods for anything special? Thanks

Reply:
Or use constructor injection ... :wink:

Reply:
I usually don't inject using fields because it makes the code untestable with unit tests without resorting to reflection. Constructor arguments sometimes also cause issues when there are cycles. :-/

Reply:
I'm using package or protected visibility for tests, not only for references.

Reply:
Hi all, thanks a lot for your valuable input. However I am not quite sure what I should do. I have not used OSGi before and I did not looked into the openhab test projects yet. So excuse my questions.
1. We are talking about a HandlerFactory here, so I do not think that constructor injection works, as the framework expects a and uses the default ctor???
2. I do not need the set and unset methods for anything special, so I can add the ```@Reference``` to the fields (was quite impressed that OSGi can do the injection to private fields). However having the unit tests in mind, I thought that these public methods would be good approach. So how do you instantiate a HandlerFactory with all its dependencies in an unit test?

I would do the following if it is ok for you all:
1. Remove set and unset methods
2. Add ```@Reference``` to the fields
3. Change visibility of these fields to package???
4. Remove the null check for ```SerialPortManager``` and check it in the BridgeHandler, so this Factory could be unit tested without the hassle of injecting these dependencies (same for ```ItemChannelLinkRegistry```)

Thanks a lot

Reply:
Sounds good. It actually doesn't matter if you call the set methods or set the fields directly in unit tests. I like shorter code so prefer the fields. Constructor injection is also viable. The buildsystem creates code to make it possible for the OSGi framework to instantiate objects.

Reply:
Yes, go for it. All three ways are valid in [OSGi Compendium Release 7](https://osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-accessing.services) and thus for us too. And each way has it pros and cons. You should use the way you are feeling comfortable with. We neither have an unofficiall guideline not an official one for its usage.

_Method injection_ may have some advantages in unit tests whereas _field injection_ leads to less code whereas _constructor injection_ fits better in combination with nullness annotations and allows `final` references.

Reply:
Hi all, I changed the way of injecting the SerialPortManager and ItemChannelLinkRegistry. Hope that it is ok now. Thanks a lot for your input 👍